### PR TITLE
imageproc things for paths refactor

### PR DESCRIPTION
**Pull request by vojtechkral**
Let me know if this makes sense & is decent enough.
I'm not sure my code style is in line with the project, I can fix this if need be...

### CODE REVIEW

File: components/site/tests/site.rs [hunk 1]
```diff
@@ -767,6 +770,91 @@ fn can_get_hash_for_static_files() {
...
+        height: 380,
+        format: Some("png".into()),
+    });
+    assert_eq!(content[1], ImageMetaResponse {
+        width: 300,
+        height: 380,
+        format: Some("jpg".into()),
+    });
+}
+
```
Comment (Keats) [inline components/site/tests/site.rs]: It all looks good to me except that - can we somehow move the tests in the imageproc module? This component already so many random tests that it's a bit hard to follow

Comment (vojtechkral) [reply components/site/tests/site.rs]: Sure! Will do :slightly_smiling_face:

Comment (vojtechkral) [reply components/site/tests/site.rs]: @Keats Moved the `imageproc` tests into its crate.
(But tbh it's kind of late over here and I should probably double-check everything tomorrow :slightly_smiling_face:)

File: components/imageproc/src/lib.rs [hunk 2]
```diff
@@ -23,9 +28,37 @@ lazy_static! {
...
+        let size = reader.into_dimensions()?;
+
+        Ok(Self { size, format })
+    }
+
+    fn is_lossy(&self) -> bool {
+        use ImageFormat::*;
+
+        // We assume lossy by default / if unknown format
+        match self.format.unwrap_or(Jpeg) {
```
Comment (Keats) [inline components/imageproc/src/lib.rs]: I think clippy will tell you to use `matches!` here

File: components/imageproc/tests/test_site.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    date_prefix: HashMap<String, json::Value>,
+}
+
+impl TestSite {
+    fn read_json<T: DeserializeOwned>(site: &Site, path: &str) -> T {
+        let content = fs::read_to_string(site.output_path.join(path)).unwrap();
+        json::from_str(&content).unwrap()
+    }
+
+    fn build() -> Self {
```
Comment (Keats) [inline components/imageproc/tests/test_site.rs]: Can we actually do the test calling only things from the imageproc module or is it too complicated?

Comment (vojtechkral) [reply components/imageproc/tests/test_site.rs]: Probably, yeah, most of it anyway. Will try to do that. Except for the date-prefix test, that one requires an actual site. Would it be ok to put that one into `site` tests?

Comment (Keats) [reply components/imageproc/tests/test_site.rs]: Why is the prefix needing a special test?

Comment (vojtechkral) [reply components/imageproc/tests/test_site.rs]: My idea was to make sure #788 is fixed, but no pressure, I don't have to include it...

Comment (Keats) [reply components/imageproc/tests/test_site.rs]: It's easy to test in the shortcode tests (and I might have already added one?) so let's leave it out there

Comment (vojtechkral) [reply components/imageproc/tests/test_site.rs]: ok :slightly_smiling_face:

Comment (Keats) [reply components/imageproc/tests/test_site.rs]: I think that + fixing CI and it's good to go!

Comment (vojtechkral) [reply components/imageproc/tests/test_site.rs]: Allright, I'll try to update later today...

Comment (vojtechkral) [reply components/imageproc/tests/test_site.rs]: Made some progress but not complete enough to update, brb tomorrow...

Comment (Keats) [reply components/imageproc/tests/test_site.rs]: Don't worry, take your time

**Comment by vojtechkral** [comment]
@Keats The PR is updated, imageproc integration tests are now site-less.
There's still a few bits which I'd like to improve, such as removing hashes from the templates image fns tests, adding a hash stability test, and improving docs, but if you'd be ok with that I'd do that as part of another PR so that we don't get mission creep in this one...

**Comment by Keats** [comment]
>  if you'd be ok with that I'd do that as part of another PR so that we don't get mission creep in this one...

Definitely, let's get all those path changes into `next` as they are starting to be chunky.

File: components/imageproc/src/lib.rs [hunk 4]
```diff
@@ -23,9 +29,34 @@ lazy_static! {
...
+impl ImageMeta {
+    fn read(path: &Path) -> ImageResult<Self> {
+        let reader = ImgReader::open(path).and_then(ImgReader::with_guessed_format)?;
+        let format = reader.format();
+        let size = reader.into_dimensions()?;
+
+        Ok(Self { size, format })
+    }
+
+    fn is_lossy(&self) -> bool {
```
Comment (Keats) [inline components/imageproc/src/lib.rs:48]: Interesting, I thought `image` would have something like that already

File: components/imageproc/src/lib.rs [hunk 5]
```diff
@@ -173,8 +235,8 @@ impl Format {
                 "png" => Some(false),
                 "gif" => Some(false),
                 "bmp" => Some(false),
-                // It is assumed that webp is lossless, but it can be both
-                "webp" => Some(false),
+                // It is assumed that webp is lossy, but it can be both
+                "webp" => Some(true),
```
Comment (Keats) [inline components/imageproc/src/lib.rs:239]: Do we still need that function with `ImageMeta::is_lossy`?

File: components/imageproc/src/lib.rs [hunk 6]
```diff
@@ -459,9 +529,89 @@ impl Processor {
...
+                .map(ImageMetaResponse::from)
+                .ok_or_else(|| Error::msg(format!("Failed to decode WebP image: {}", path.display())))
+        }
+        _ => ImageMeta::read(path).map(ImageMetaResponse::from).map_err(|e| error(e.into())),
+    }
+}
+
+/// Assert that `address` matches `prefix` + RESIZED_FILENAME regex + "." + `extension`,
+/// this is useful in test so that we don't need to hardcode hash, which is annoying.
+pub fn assert_processed_path_matches(path: &str, prefix: &str, extension: &str) {
```
Comment (Keats) [inline components/imageproc/src/lib.rs:603]: can that be moved to tests or with a cfg[test]?

**Review by Keats — APPROVED** [review]
I like what I see! I'll merge that now to avoid staleness

File: components/imageproc/src/lib.rs (hunk 6 above)
Comment (vojtechkral) [reply components/imageproc/src/lib.rs:603]: `cfg[test]` can't be used here as integration tests build against a non-test library, iirc. But it could be moved to tests maybe, yeah, I'll try...
